File: sdk/node/src/transport.ts
import type { ObservrEvent } from "./types.js";

const BATCH_SIZE = 50;
const FLUSH_INTERVAL_MS = 1000;

export class Transport {
  private readonly collectorUrl: string;
  private readonly service: string;
  private queue: ObservrEvent[] = [];
  private timer: ReturnType<typeof setInterval> | null = null;
  private stopped = false;

  constructor(collectorUrl: string, service: string) {
    this.collectorUrl = collectorUrl.replace(/\/$/, "");
    this.service = service;
    this.timer = setInterval(() => this._flush(), FLUSH_INTERVAL_MS);
    // Allow process to exit without waiting for timer
    if (this.timer.unref) this.timer.unref();
  }

  send(event: Omit<ObservrEvent, "service"> & { service?: string }): void {
    if (this.stopped) return;
    const full: ObservrEvent = { service: this.service, ...event };
    if (this.queue.length >= 10_000) return; // drop silently
    this.queue.push(full);
    if (this.queue.length >= BATCH_SIZE) this._flush();
  }

  async flush(timeoutMs = 5000): Promise<void> {
    const deadline = Date.now() + timeoutMs;
    while (this.queue.length > 0 && Date.now() < deadline) {
      await this._flush();
      if (this.queue.length > 0) await sleep(50);
    }
  }

  async shutdown(): Promise<void> {
    this.stopped = true;
    if (this.timer) {
      clearInterval(this.timer);
      this.timer = null;
    }
    await this.flush();
  }

  private _flush(): Promise<void> {
    if (this.queue.length === 0) return Promise.resolve();
    const batch = this.queue.splice(0, BATCH_SIZE);
    return this._postBatch(batch);
  }

  private async _postBatch(batch: ObservrEvent[]): Promise<void> {
    try {
      const url = `${this.collectorUrl}/events`;
      await fetch(url, {
        method: "POST",
        headers: { "Content-Type": "application/json" },
        body: JSON.stringify({ events: batch }),
        signal: AbortSignal.timeout(3000),
      });
    } catch {
      // Collector not running — silently discard
    }
  }
}

function sleep(ms: number): Promise<void> {
  return new Promise((r) => setTimeout(r, ms));
}

File: sdk/go/transport.go
package observr

import (
	"bytes"
	"encoding/json"
	"net/http"
	"time"
)

const (
	queueSize     = 10_000
	flushInterval = time.Second
)

// Transport batches events and POSTs them to the collector.
type Transport struct {
	url    string
	queue  chan map[string]any
	done   chan struct{}
	client *http.Client
}

// NewTransport creates a Transport targeting collectorURL.
func NewTransport(collectorURL string) *Transport {
	return &Transport{
		url:    collectorURL + "/events",
		queue:  make(chan map[string]any, queueSize),
		done:   make(chan struct{}),
		client: &http.Client{Timeout: 5 * time.Second},
	}
}

// Start begins the background flush goroutine.
func (t *Transport) Start() {
	go t.loop()
}

// Enqueue adds an event to the send queue. Drops silently if full.
func (t *Transport) Enqueue(event map[string]any) {
	select {
	case t.queue <- event:
	default:
	}
}

// Shutdown flushes remaining events and stops the goroutine.
func (t *Transport) Shutdown() {
	close(t.done)
	var batch []map[string]any
	for {
		select {
		case e := <-t.queue:
			batch = append(batch, e)
		default:
			if len(batch) > 0 {
				t.flush(batch)
			}
			return
		}
	}
}

func (t *Transport) loop() {
	ticker := time.NewTicker(flushInterval)
	defer ticker.Stop()
	var batch []map[string]any
	for {
		select {
		case e := <-t.queue:
			batch = append(batch, e)
		case <-ticker.C:
			if len(batch) > 0 {
				t.flush(batch)
				batch = nil
			}
		case <-t.done:
			return
		}
	}
}

func (t *Transport) flush(events []map[string]any) {
	body, err := json.Marshal(events)
	if err != nil {
		return
	}
	req, err := http.NewRequest(http.MethodPost, t.url, bytes.NewReader(body))
	if err != nil {
		return
	}
	req.Header.Set("Content-Type", "application/json")
	resp, err := t.client.Do(req)
	if err != nil {
		return
	}
	resp.Body.Close()
}
